activerecord/lib/active_record/migration.rb

```diff
@@ -563,17 +563,16 @@ def initialize(app, file_watcher: ActiveSupport::FileUpdateChecker)
         @app = app
         @needs_check = true
         @mutex = Mutex.new
-        @file_watcher = file_watcher
+
+        @watcher = build_watcher(file_watcher) do
+          @needs_check = true
+          ActiveRecord::Migration.check_pending!(connection)
+          @needs_check = false
+        end
       end
 
       def call(env)
         @mutex.synchronize do
-          @watcher ||= build_watcher do
-            @needs_check = true
-            ActiveRecord::Migration.check_pending!(connection)
-            @needs_check = false
-          end
-
           if @needs_check
             @watcher.execute
```

Comment:
This change doesn't seem like it'll fix the problem since `@watcher.execute` will invoke the block passed to the watcher's constructor (now lines 568-570).

That means that it's still possible to perform autoloading while the `@mutex` in this class in still held, which is what caused the original deadlock.

I think this change makes it so that deadlocks are less likely to happen on startup, which is an improvement, but it doesn't make the deadlocks impossible.

For a deadlock to occur now, we'd have to start the server, add/edit a migration file so that `@needs_check` becomes true, then hammer the web server with concurrent requests. One thread will acquire the mutex and run `@water.execute`, which will try to get a database connection and perform some autoloading, while another thread might be stuck trying to get the mutex.

Reply:
> which will try to get a database connection and perform some autoloading

When the watcher is built, it would already have checked out a connection and result in an autoload. Even if `@watcher.execute` is called where a connection is checked out again, whatever that needs to be autoloaded should have been loaded.